swatch-core/src/main/java/org/candlepin/subscriptions/db/EventRecordRepository.java

```diff
@@ -158,4 +164,46 @@ Stream<EventRecord> fetchOrderedEventStream(
       @Param("orgId") String orgId,
       @Param("serviceType") String serviceType,
       @Param("after") OffsetDateTime after);
+
+  /**
+   * Find all {@link EventRecord}s that share the same lookup key.
+   *
+   * <pre>
+   *   NOTE: This method builds a set of tuples from the supplied keys and executes
+   *         a native Postgres query. This is not possible in JPA.
+   * </pre>
+   *
+   * @param keys the {@link EventKey} to match on.
+   * @return a list of conflicting events
+   */
+  default List<EventRecord> findConflictingEvents(List<EventKey> keys) {
```

Comment:
My main concern here is not about locating huge data in memory since I don't expect too many conflicting events, but about the performance and the number of memory allocations.
With the solution in place, we're doing:
1.- `List<Event> eventsToResolve`: keep in memory the batch of events we receive from the topic or the admin API (let's say we process 500 events)
2.- `Map<EventKey, List<EventRecord>> allConflicting`: run a single query to find all the events by event key (I don't expect too much data, but in the worst scenario, we might need to allocate unlimited events by event key which is a risk). 
3.- `List<EventRecord> resolvedEvents`: the actual events to be persisted. 
4.- loop over the `eventsToResolve` list and for every item:
4.1.- create again the `EventKey` class
4.2.- check whether the event key exists in `allConflicting`
4.2.1.1.- if not exists, just add the event into `resolvedEvents`
4.2.2.1.- if exists, `Map<String, Double> deductions`: aggregate the metrics with the same event key, 
4.2.2.2.- if no `deductions`, just add the event into `resolvedEvents`
4.2.2.3.- if there are `deductions`, add the new event with the deduction for the metric into `resolvedEvents`
4.2.2.4.- if the event has other metrics with no deductions, add new events only for these metrics into `resolvedEvents`

Note that from the steps 1, 2, 3; we're keeping almost the same data duplicated in memory, plus the performance penalty of creating the event key. 

I suggested changing the above by:

1.- as suggested by another comment, we can directly use `Map<EventKey, List<Event>> eventsToResolve` as input
2.- instead of using `Map<EventKey, List<EventRecord>> allConflicting` which can keep unlimited data in memory, modify the query to return `Map<EventKey, Map<String, Double>> eventsByDeduction`. Note that I tried to do this and it worked fine with something like:

```sql
select org_id, event_type, event_source, instance_id, timestamp, measurements->>'uom' as uom, sum(cast(measurements->>'value' as double precision)) as total_value
from (
select org_id, event_type, event_source, instance_id, timestamp, jsonb_array_elements(data->'measurements') as measurements
from events
where (org_id, event_type, event_source, instance_id, timestamp) in (%s)
) a
group by org_id, event_type, event_source, instance_id, timestamp, measurements->>'uom'
```

This way, the logic within the step 4. is much simpler and performant since we don't need to perform any aggregation (the database is doing it for us).

Reply:
I chatted with Kevin about this suggestion a little while ago and we both agreed that at some point we may need to a little more than just aggregation with conflicting events (i.e event_source prioritization). This is a valid concern, though, I'm not sure it would be very common to have many conflicting events. I DO think it would be nice to have this solution in our back pocket should we need it.

I'm happy to have a chat if you feel strongly for this change though.